### ISSUE TITLE
Using the modulo operator % as string formatter

### ISSUE THREAD

**Issue by phorward** [QQuick/Transcrypt#467]
Hi there,

why does this piece of code
```python
test = "get"
key = "test"
print(test, key)
print("_%s%s%s" % (test, key[0].upper(), key[1:]))
```
result in 
```
get test
NaN
```
in the JavaScript console, althought I did not use --jsmod to enforce JavaScript modulo support.
Desired output should be
```
get test
_getTest
```
Is it a bug or a desired behavior?

**Comment by neriusmika** [comment]
Documentation:
(9.2) String formatting is done through the format method or through f-strings, as this is considered most flexible. Using % for formatting isn't supported. There's currently no 'string formatting mini language'. All of JavaScript's string formatting facilities are supported.

**Comment by phorward** [comment]
Thanks for the hint... maybe it's better for me to first read the complete documentation. I'm trying to port a huger library (https://github.com/viur-framework/html5) from PyJS to Transcrypt and there are many pitfalls to deal with. Thanks again, and now its time to read the docs first... (its just "RTFM" for me ;-))

**Comment by JdeH** [comment]
Indeed if porting something large, reading the docs will probably pay of, although they may be a bit terse.
There are quite some switches, pragma's and other stuff that may make your life easier.

**Comment by neriusmika** [comment]
BTW, I almost finished python style string formatting, maybe we can adapt it for % style formatting too. But there is about 300 lines of non optimized js code. So Jacques will need to decide to replace the old format function or to make it available via pragma.

**Comment by JdeH** [comment]
I am in favor of making it available, but not as default, since it's quite large.
This isn't a problem, I expect we can handle it much like the -xt switch.
Only here the default will be: not available (just like e.g. complex number support).
As an option I consider this a very useful addition.

**Comment by neriusmika** [comment]
Jacques, I made string formatting in my fork, you can add it to transcrypt if you want. % formatting is not working but I don't think it would be hard to implement in the future. In the current implementation there is support for builtin format(), str.format(), custom formatting via \_\_format\_\_ method.

P.S. available via -sf switch.

**Comment by JdeH** [comment]
I am reviewing your changes and probably will have to ask a few things.
In the following code, why is aClass2 needed? (see ??? in code).

```
    var issubclass = function (aClass, classinfo) {
        if (classinfo instanceof Array) {   // Assume in most cases it isn't, then making it recursive rather than two functions saves a call
__pragma__ ('ifdef', '__esv6__')
            for (let aClass2 of classinfo) {
__pragma__ ('else')
            for (var index = 0; index < classinfo.length; index++) {
                var aClass2 = classinfo [index];
__pragma__ ('endif')
                if (issubclass (aClass, aClass2)) {
                    return true;
                }
            }
            return false;
        }
        try {
            var aClass2 = aClass;               # ??? Why is aClass2 needed? You say for safety, but I don't see any danger...
            if (aClass2 == classinfo)
                return true;
            else {
                var bases = [].slice.call (aClass2.__bases__);
                while (bases.length) {
                    aClass2 = bases.shift ();
                    if (aClass2 == classinfo)
                        return true;
                    if (aClass2.__bases__.length)
                        bases = [].slice.call (aClass2.__bases__).concat (bases);
                }
                return false;
            }
        }
        catch (exception) {     // Using issubclass on primitives assumed rare 
            return aClass == classinfo || classinfo == object;
        }
    };
    __all__.issubclass = issubclass;

```

**Comment by JdeH** [comment]
I am reviewing your changes and have some questions.
I will pose them as reaction to your individual commits.

KR
J

**Comment by neriusmika** [comment]
In the latest commit I added support for builtin ascii. It's quite useful function and is needed by string formatting. But to work correctly, page charset must be utf-8, so I modified autotest to produce a header entry for that. And by the way, without this charset all strings in javascript are treated as non unicode.

P.S. found a bug in a string formatting when argument is null or undefined. Probably easy to fix.
